File: rs/src/l2_rotary_lock2.rs
#![allow(non_snake_case)]

use std::cmp;
use std::hash::{Hash, Hasher};

#[derive(Eq)]
struct Dials
{
    dial1: i32,
    dial2: i32,
}

impl PartialEq for Dials {
    fn eq(&self, other: &Self) -> bool {
        return self.dial1 == other.dial1 && self.dial2 == other.dial2;
    }
}

impl Hash for Dials {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.dial1.hash(state);
        self.dial2.hash(state);
    }
}

type SolutionsT = std::collections::HashMap<Dials, i64>;

fn get_distance(target: i32, position: i32, N: i32) -> i32 {
    let mut positive_move = (target - position) % N;
    if positive_move < 0 {
        positive_move += N;
    }
    let negative_move = N - positive_move;  // positive number
    return cmp::min(positive_move, negative_move);
}

fn insert_solution(new_solutions: &mut SolutionsT, N: i32, target: i32, dial1: i32, dial2: i32, distance: i64)
{
    let new_distance = distance + get_distance(target, dial1, N) as i64;
    let key = Dials{ dial1: cmp::min(dial2, target), dial2: cmp::max(dial2, target) };
    let it = new_solutions.get(&key);
    let value = it.unwrap_or(&i64::MAX);
    new_solutions.insert(key, cmp::min(*value, new_distance));
}

pub fn getMinCodeEntryTime(N: i32, _M: i32, C: &Vec<i32>) -> i64 {
    if C.is_empty() {
        return 0;
    }
    let max_value = i64::MAX;
    let mut solutions = SolutionsT::default();
    solutions.insert(Dials{ dial1: 1, dial2: 1 }, 0);
    for &target in C {
        let mut new_solutions = SolutionsT::default();
        for (dials, &distance) in &solutions {
            // we turn dial1
            insert_solution(&mut new_solutions, N, target, dials.dial1, dials.dial2, distance);
            // we turn dial2
            insert_solution(&mut new_solutions, N, target, dials.dial2, dials.dial1, distance);
        }
        solutions = new_solutions;
    }
    let mut min_distance = max_value;
    for (_, &distance) in &solutions {
        min_distance = cmp::min(min_distance, distance);
    }
    return min_distance;
}

File: js/l2_rotary_lock2.js
// This was converted from Python to Javascript using https://www.codeconvert.ai/

function getDistance(target, position, N) {
    // code convert does not deal with modulo the same way as the reference code
    //const positiveMove = (target - position) % N;
    let positiveMove = 0;
    if (target > position) {
        positiveMove = (target - position) % N;
    } else {
        positiveMove = (position - target) % N;
    }
    const negativeMove = N - positiveMove; // positive number
    return Math.min(positiveMove, negativeMove);
}

function getMinCodeEntryTime(N, M, C) {
    // Constraints:
    //   3 ≤ N ≤ 1,000,000,000   N is the number of integers
    //   1 ≤ M ≤ 3,000           M is the number of locks
    //   1 ≤ Ci ≤ N              Ci is the combination
    // Complexity: O(M^2)

    if (C.length === 0) {
        return 0;
    }
    const solutions = new Map();
    solutions.set(JSON.stringify([1, 1]), 0);
    
    for (let i = 0; i < Math.min(M, C.length); i++) { // M iterations
        const target = C[i];
        const newSolutions = new Map();
        
        for (const [key, distance] of solutions) { // 2*M iterations at most
            const [dial1, dial2] = JSON.parse(key);
            
            // we turn dial1
            const distance1 = distance + getDistance(target, dial1, N);
            const key1 = JSON.stringify([Math.min(dial2, target), Math.max(dial2, target)]);
            newSolutions.set(key1, Math.min(newSolutions.get(key1) || Infinity, distance1));
            
            // we turn dial2
            const distance2 = distance + getDistance(target, dial2, N);
            const key2 = JSON.stringify([Math.min(dial1, target), Math.max(dial1, target)]);
            newSolutions.set(key2, Math.min(newSolutions.get(key2) || Infinity, distance2));
        }
        solutions.clear();
        for (const [key, value] of newSolutions) {
            solutions.set(key, value);
        }
    }
    return Math.min(...Array.from(solutions.values()));
}

function tests() {
    const fn = (N, C) => [N, C.length, C];
    const metaCases = ["meta", [
        [[3, [1, 2, 3]], 2],
        [[10, [9, 4, 4, 8]], 6],
    ]];
    const extra1Cases = ["extra1", [
        [[0, []], 0],
        [[3, []], 0],
        [[10, []], 0],
        [[10, [4]], 3],
        [[10, [9]], 2],
        [[10, [9, 9, 9, 9]], 2],
    ]];
    const extra2Cases = ["extra2", [
        [[10, [6, 2, 4, 8]], 10],
        [[10, [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]], 9],
        [[4, [4, 3, 2, 1, 2, 3, 4]], 5],
    ]];
    return [getMinCodeEntryTime, fn, [metaCases, extra1Cases, extra2Cases]];
}

module.exports = {tests};
